script/cache_critical_dns

```diff
@@ -122,50 +249,83 @@ end
   end
 end.compact
 
-def loop
+def run(dns_client)
```

Comment:
This is more complicated than I expected. :)

```ruby
def nilempty(v)
  # lol ruby
  if v.nil?
    nil
  elsif v.method_defined?(:empty?) && v.empty?
    nil
  else
    v
  end
end

def env_srv_name(env_name)
  nilempty(ENV["#{env_name}_SRV"])
end

def run(dns_client)
  resolved = {}  # {hostname: [address, ...]}
  errors = Hash.new(0)

  @unresolved_host_env_vars.each do |var|
    name = ENV[var]

    HOST_RESOLVER_CACHE[var] ||= ResolverCache.new(begin
      if (srv_name = env_srv_name(var))
        SRVName.new(srv_name)
      else
        Name.new(name)
      end
    end)

    HOST_HEALTHY_CACHE[var] ||= HealthyCache.new(HOST_RESOLVER_CACHE[var], HEALTH_CHECKS[var.to_sym])

    address = begin
      HOST_HEALTHY_CACHE[var].first_healthy(dns_client)
    rescue => e
      error("#{var}: #{name}: #{e}")
    end
    if address.nil?
      error("#{var}: #{name}: no address")
      errors[name] += 1
      next
    end

    resolved[name] = [address]
  end

  # hosts(5) faff here ...
end
```

(My Ruby sucks and I have not tested to see whether this parses; please forgive any dumb mistakes.)

Some caveats:

* This will enable health checks on the legacy path.  This should be harmless -- if the postgres/redis gems are available.  I can't remember what the deal is with those.
* Behaviour may change in some weird way for any FOSS users with names that map to multiple addresses.  Should not be a problem with functional health checks, though.

Reply:
I think adding health checks even for non `_SRV` hosts is a big improvement. I've made incorporated these changes and it makes the run loop much simpler. Thanks for the feedback.